// plugin-welcome/src/api/models/EntSubjectEdges.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntLessonplan,
    EntLessonplanFromJSON,
    EntLessonplanFromJSONTyped,
    EntLessonplanToJSON,
    EntTeacher,
    EntTeacherFromJSON,
    EntTeacherFromJSONTyped,
    EntTeacherToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntSubjectEdges
 */
export interface EntSubjectEdges {
    /**
     * 
     * @type {EntTeacher}
     * @memberof EntSubjectEdges
     */
    professorID?: EntTeacher;
    /**
     * SubjectID holds the value of the Subject_ID edge.
     * @type {Array<EntLessonplan>}
     * @memberof EntSubjectEdges
     */
    subjectID?: Array<EntLessonplan>;
}

export function EntSubjectEdgesFromJSON(json: any): EntSubjectEdges {
    return EntSubjectEdgesFromJSONTyped(json, false);
}

export function EntSubjectEdgesFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntSubjectEdges {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'professorID': !exists(json, 'professorID') ? undefined : EntTeacherFromJSON(json['professorID']),
        'subjectID': !exists(json, 'subjectID') ? undefined : ((json['subjectID'] as Array<any>).map(EntLessonplanFromJSON)),
    };
}

export function EntSubjectEdgesToJSON(value?: EntSubjectEdges | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'professorID': EntTeacherToJSON(value.professorID),
        'subjectID': value.subjectID === undefined ? undefined : ((value.subjectID as Array<any>).map(EntLessonplanToJSON)),
    };
}
