// plugin-welcome/src/api/models/EntSubject.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntSubjectEdges,
    EntSubjectEdgesFromJSON,
    EntSubjectEdgesFromJSONTyped,
    EntSubjectEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntSubject
 */
export interface EntSubject {
    /**
     * Credit holds the value of the "Credit" field.
     * @type {number}
     * @memberof EntSubject
     */
    credit?: number;
    /**
     * Price holds the value of the "Price" field.
     * @type {number}
     * @memberof EntSubject
     */
    price?: number;
    /**
     * SubjectName holds the value of the "Subject_Name" field.
     * @type {string}
     * @memberof EntSubject
     */
    subjectName?: string;
    /**
     * 
     * @type {EntSubjectEdges}
     * @memberof EntSubject
     */
    edges?: EntSubjectEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntSubject
     */
    id?: number;
}

export function EntSubjectFromJSON(json: any): EntSubject {
    return EntSubjectFromJSONTyped(json, false);
}

export function EntSubjectFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntSubject {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'credit': !exists(json, 'Credit') ? undefined : json['Credit'],
        'price': !exists(json, 'Price') ? undefined : json['Price'],
        'subjectName': !exists(json, 'Subject_Name') ? undefined : json['Subject_Name'],
        'edges': !exists(json, 'edges') ? undefined : EntSubjectEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntSubjectToJSON(value?: EntSubject | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'Credit': value.credit,
        'Price': value.price,
        'Subject_Name': value.subjectName,
        'edges': EntSubjectEdgesToJSON(value.edges),
        'id': value.id,
    };
}
