// plugin-welcome/src/api/models/EntTeacherEdges.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntLessonplan,
    EntLessonplanFromJSON,
    EntLessonplanFromJSONTyped,
    EntLessonplanToJSON,
    EntSubject,
    EntSubjectFromJSON,
    EntSubjectFromJSONTyped,
    EntSubjectToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntTeacherEdges
 */
export interface EntTeacherEdges {
    /**
     * TeacherID holds the value of the Teacher_ID edge.
     * @type {Array<EntSubject>}
     * @memberof EntTeacherEdges
     */
    teacherID?: Array<EntSubject>;
    /**
     * TeachersID holds the value of the Teachers_ID edge.
     * @type {Array<EntLessonplan>}
     * @memberof EntTeacherEdges
     */
    teachersID?: Array<EntLessonplan>;
}

export function EntTeacherEdgesFromJSON(json: any): EntTeacherEdges {
    return EntTeacherEdgesFromJSONTyped(json, false);
}

export function EntTeacherEdgesFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntTeacherEdges {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'teacherID': !exists(json, 'teacherID') ? undefined : ((json['teacherID'] as Array<any>).map(EntSubjectFromJSON)),
        'teachersID': !exists(json, 'teachersID') ? undefined : ((json['teachersID'] as Array<any>).map(EntLessonplanFromJSON)),
    };
}

export function EntTeacherEdgesToJSON(value?: EntTeacherEdges | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'teacherID': value.teacherID === undefined ? undefined : ((value.teacherID as Array<any>).map(EntSubjectToJSON)),
        'teachersID': value.teachersID === undefined ? undefined : ((value.teachersID as Array<any>).map(EntLessonplanToJSON)),
    };
}
