// @backstage/cli/dist/cjs/dev-c1dbad4d.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

require('commander');
require('chalk');
var index = require('./index-8e863345.js');
require('fs-extra');
require('@backstage/cli-common');
var configLoader = require('@backstage/config-loader');
var config = require('@backstage/config');
var webpack2 = _interopDefault(require('webpack'));
require('path');
require('fork-ts-checker-webpack-plugin');
require('html-webpack-plugin');
require('react-dev-utils/ModuleScopePlugin');
require('start-server-webpack-plugin');
require('webpack-node-externals');
var paths = require('./paths-90fb4262.js');
require('mini-css-extract-plugin');
require('./svgrTemplate-2930b0e1.js');
require('child_process');
require('util');
require('./run-76e92ad7.js');

async function serveBackend(options) {
  const paths2 = paths.resolveBundlingPaths(options);
  const config2 = await paths.createBackendConfig(paths2, {
    ...options,
    isDev: true
  });
  const compiler = webpack2(config2);
  const watcher = compiler.watch({
    poll: true
  }, (err) => {
    if (err) {
      console.error(err);
    } else
      console.log("Build succeeded");
  });
  const waitForExit = async () => {
    for (const signal of ["SIGINT", "SIGTERM"]) {
      process.on(signal, () => {
        watcher.close(() => console.log("Stopped watcher"));
        process.exit();
      });
    }
    return new Promise(() => {
    });
  };
  return waitForExit;
}

var dev = async (cmd) => {
  var _a;
  const appConfigs = await configLoader.loadConfig({
    env: (_a = process.env.NODE_ENV) != null ? _a : "development",
    rootPaths: [index.paths.targetRoot, index.paths.targetDir]
  });
  const waitForExit = await serveBackend({
    entry: "src/index",
    checksEnabled: cmd.check,
    inspectEnabled: cmd.inspect,
    config: config.ConfigReader.fromConfigs(appConfigs),
    appConfigs
  });
  await waitForExit();
};

exports.default = dev;
